Sources/TensorFlow/Core/LazyTensorTFFunctionBuilder.swift

```diff
@@ -0,0 +1,324 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import CTensorFlow
+
+/// Swift-side convenience wrapper for a `TF_Graph*`. It holds a pointer to the underlying
+/// TF_Graph and also exposes the some aspects of the TF_Graph as properties.
+class TFGraph {
+    /// The `TF_Operation *` type.
+    typealias CTFOperation = OpaquePointer
+
+    enum Input {
+        case single(TF_Output)
+        case list([TF_Output])
+    }
+
+    /// The pointer to the underlying TF Graph.
+    let cTFGraph: CTFGraph = TF_NewGraph()
+    var inputs: [TF_Output] = []
+    var nodes: [CTFOperation?] = []
+    var outputs: [TF_Output] = []
+    /// An array representing how the outputs are grouped. The grouping
+    /// corresponds to a higher-level notion like TensorGroup.
+    var outputGroupCounts: [Int] = []
+    var name: String { "lazyTrace_\(nodeCounter)" }
+
+    /// A status object to pass to TF graph building operations.
+    private let status: CTFStatus = TF_NewStatus()
+
+    /// Counter that is used for number the generated graph nodes.
+    private var nodeCounter: Int = 0
+
+    init(_ lazyTrace: LazyTensorTrace) {
+        var nodesCache: [ObjectIdentifier: CTFOperation?] = [:]
+        for op in lazyTrace.operations {
+            let opInputs = op.inputs.map { input -> TFGraph.Input in
+                switch input {
+                case LazyTensorOperation.Input.single(let h):
+                    return TFGraph.Input.single(makeTFOutput(h, nodesCache))
+                case LazyTensorOperation.Input.list(let elements):
+                    let tfInputs = elements.map { makeTFOutput($0, nodesCache) }
+                    return TFGraph.Input.list(tfInputs)
+                }
+            }
+            let graphNode = makeTFGraphNode(
+                name: op.name,
+                attrs: op.attributes,
+                inputs: opInputs,
+                device: op.deviceName)
+            nodesCache[ObjectIdentifier(op)] = graphNode
+            if op.name != "Placeholder" { nodes.append(graphNode) }
+        }
+        self.inputs = lazyTrace.inputs.map {
+            TF_Output(oper: nodesCache[ObjectIdentifier($0)]!, index: 0)
+        }
+        for output in lazyTrace.outputs {
+            let graphNode = nodesCache[ObjectIdentifier(output)]!
+            outputGroupCounts.append(output.outputCount)
+            outputs += Array((0..<output.outputCount).map {
+                    TF_Output(oper: graphNode, index: Int32($0))
+                })
```

Comment:
`TF_Output(oper: graphNode, index: Int32($0))` and `})` should be indented by 4 spaces less each.